File: Editorial/1102_Path_With_Maximum_Minimum_Value/CPP/1_Iteration_BFS.cpp
class Solution {
    // 4 directions to a cell's possible neighbors.
    vector<vector<int>> dirs{{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

public:
    int maximumMinimumPath(vector<vector<int>>& grid) {
        int R = grid.size(), C = grid[0].size();
        int curScore = min(grid[0][0], grid[R - 1][C - 1]);

        // Start with "curScore", check if we can find a path having score
        // equals "curScore". If so, return curScore as the maximum score,
        // otherwise, decrease curScore by 1 and repeat these steps.
        while (curScore >= 0) {
            if (pathExists(grid, curScore)) {
                return curScore;
            } else {
                curScore = curScore - 1;
            }
        }
        return -1;
    }

    // Check if we can find a path of score equals curScore.
    bool pathExists(vector<vector<int>>& grid, int curScore) {
        int R = grid.size(), C = grid[0].size();
        // Maintain the visited status of each cell. Initialize the status of
        // all the cells as false (unvisited).
        vector<vector<bool>> visited(R, vector<bool>(C, false));

        // Put the starting cell grid[0][0] in the deque and mark it as visited.
        visited[0][0] = true;
        deque<pair<int, int>> dq({{0, 0}});

        auto push = [&](int row, int col) {
            if (row == -1 || col == -1 || row == R || col == C ||
                visited[row][col] || grid[row][col] < curScore)
                return;
            dq.push_back({row, col});
            visited[row][col] = true;
        };

        // While we still have cells of value larger than or equal to curScore.
        while (!dq.empty()) {
            int curRow = dq.front().first, curCol = dq.front().second;
            dq.pop_front();

            // If the current cell is the bottom-right cell, return true.
            if (curRow == R - 1 && curCol == C - 1) {
                return true;
            }

            // Check if the current cell has any unvisited neighbors with value
            // larger than or equal to curScore. If so, we put this neighbor to
            // the deque and mark it as visited
            push(curRow + 1, curCol);
            push(curRow - 1, curCol);
            push(curRow, curCol + 1);
            push(curRow, curCol - 1);
        }
        return false;
    }
};

File: Editorial/1101-1200/1102_Path_With_Maximum_Minimum_Value/Python3/1_Iteration_BFS.py
class Solution:
    def maximumMinimumPath(self, grid: List[List[int]]) -> int:
        R = len(grid)
        C = len(grid[0])

        # 4 directions to a cell's possible neighbors.
        dirs = [(1, 0), (0, 1), (-1, 0), (0, -1)]

        # Check if we can find a path of score equals val.
        def path_exists(cur_score):

            # Maintain the visited status of each cell. Initalize the status of
            # all the cells as false (unvisited).
            visited = [[False] * C for _ in range(R)]
            visited[0][0] = True

            # Put the starting cell grid[0][0] in the deque and mark it as visited.
            dq = collections.deque([(0, 0)])

            # While we still have cells of value larger than or equal to val.
            while dq:
                cur_row, cur_col = dq.popleft()

                # If the current cell is the bottom-right cell, return true.
                if cur_row == R - 1 and cur_col == C - 1:
                    return True
                for d_row, d_col in dirs:
                    new_row = cur_row + d_row
                    new_col = cur_col + d_col

                    # Make sure (new_row, new_col) is on the grid.
                    if not (0 <= new_row < R and 0 <= new_col < C):
                        continue

                    # Check if the current cell has any unvisited neighbors
                    # with value greater than or equal to cur_score.
                    if (
                        grid[new_row][new_col] >= cur_score
                        and not visited[new_row][new_col]
                    ):
                        # If so, we put this neighbor to the deque and mark it as visited
                        visited[new_row][new_col] = True
                        dq.append((new_row, new_col))
            return False

        cur_score = min(grid[0][0], grid[R - 1][C - 1])

        # Start with "cur_score", check if we can find a path of score equals cur_score.
        # If so, return cur_score as the maximum score, otherwise, decrease cur_score
        # by 1 and repeat these steps.
        while cur_score >= 0:
            if path_exists(cur_score):
                return cur_score
            else:
                cur_score -= 1
